txcache/selectionTracker_test.go

```diff
@@ -0,0 +1,252 @@
+package txcache
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/multiversx/mx-chain-core-go/data/block"
+	"github.com/stretchr/testify/require"
+)
+
+func createMockedHeaders(numOfHeaders int) []*block.Header {
+	headers := make([]*block.Header, 0)
+
+	for i := 0; i < numOfHeaders; i++ {
+		headers = append(headers, &block.Header{
+			Nonce:    uint64(i),
+			PrevHash: []byte(fmt.Sprintf("prevHash%d", i)),
+			RootHash: []byte(fmt.Sprintf("rootHash%d", i)),
+		})
+	}
+
+	return headers
+}
+
+func proposeBlocks(t *testing.T, numOfBlocks int, selectionTracker *selectionTracker, headers []*block.Header) {
+	wg := sync.WaitGroup{}
+	wg.Add(numOfBlocks)
+
+	for i := 0; i < numOfBlocks; i++ {
+		go func(index int) {
+			err := selectionTracker.OnProposedBlock([]byte(fmt.Sprintf("blockHash%d", i)), &block.Body{}, headers[index])
+			require.Nil(t, err)
+			defer wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+}
+
+func executeBlocks(t *testing.T, numOfBlocks int, selectionTracker *selectionTracker, headers []*block.Header) {
+	wg := sync.WaitGroup{}
+	wg.Add(numOfBlocks)
+
+	for i := 0; i < numOfBlocks; i++ {
+		go func(index int) {
+			err := selectionTracker.OnExecutedBlock(headers[index])
+			require.Nil(t, err)
+			defer wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+}
+
+func TestNewSelectionTracker(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should error", func(t *testing.T) {
+		t.Parallel()
+
+		selTracker, err := NewSelectionTracker(nil)
+		require.Nil(t, selTracker)
+		require.Error(t, err)
+	})
+
+	t.Run("should work", func(t *testing.T) {
+		t.Parallel()
+
+		txCache := newCacheToTest(maxNumBytesPerSenderUpperBoundTest, 3)
+		_, err := NewSelectionTracker(txCache)
+		require.Nil(t, err)
+	})
+}
+
+func TestSelectionTracker_OnProposedBlockShouldErr(t *testing.T) {
+	t.Parallel()
+
+	txCache := newCacheToTest(maxNumBytesPerSenderUpperBoundTest, 3)
+	selTracker, err := NewSelectionTracker(txCache)
+	require.Nil(t, err)
+
+	t.Run("should err nil block hash", func(t *testing.T) {
+		t.Parallel()
+
+		err = selTracker.OnProposedBlock(nil, nil, nil)
+		require.Equal(t, err, errNilBlockHash)
+	})
+
+	t.Run("should err nil header", func(t *testing.T) {
+		t.Parallel()
+
+		err = selTracker.OnProposedBlock([]byte("hash1"), nil, nil)
+		require.Equal(t, err, errNilBlockBody)
+	})
+
+	t.Run("should err nil header", func(t *testing.T) {
+		t.Parallel()
+
+		blockBody := block.Body{}
+		err = selTracker.OnProposedBlock([]byte("hash1"), &blockBody, nil)
+		require.Equal(t, err, errNilHeaderHandler)
+	})
+}
+
+func TestSelectionTracker_OnProposedBlockShouldWork(t *testing.T) {
+	t.Parallel()
+
+	txCache := newCacheToTest(maxNumBytesPerSenderUpperBoundTest, 3)
+	selTracker, err := NewSelectionTracker(txCache)
+	require.Nil(t, err)
+
+	numOfBlocks := 20
+	headers := createMockedHeaders(numOfBlocks)
+	proposeBlocks(t, numOfBlocks, selTracker, headers)
+}
```

Comment:
Check the object at the end of the test? E.g. maybe check length of the `blocks` (state) field?